Refactor ESLint and Babel Project Configurations

bendvc opened the pull request:
# Update 

A lot of the description of this PR's solution below is out of date.

# Special Notes

To get eslinting working on your IDE you have to set your working directory setting, below is an example of vscode 

```
// User/settings.json

"eslint.workingDirectories": [
    {"mode": "auto"}
]

```
<!--- Provide a short summary of your changes in the Title field above -->

# Description

It was reported that IDE ESLint integrations was no longer working for mono repo developers (salesforce developers). Although developing individual generated projects worked fine. This in itself wasn't a big issue as we (salesforce) developers could still run lint commands via the CLI and CI integration was still working. This issue stemmed from the fact that all the eslint configuration was defined in the `pwa-kit-cli` package which wasn't in the npm resolution path. Meaning if you edit a template, your IDE would never find either `eslint` or its configuration.

This raised a possible future problem that making ESLint IDE integration work correctly would be a breaking change. So it would be smart to figure this out before releasing V2 of the PWA-Kit. 

# Solution Explanation

This PR attempts to to solve this problem by defining 2 new packages `eslint-config-pwa-kit` and `babel-preset-pwa-kit` and defining boundaries for what projects/packages get linted by what rule set. Creating these packages allows us to specify all the _eslint_ and babel dependencies in their own packages respectively and are directly use by the templates. We also integrate _eslint_ on the mono repo level, to handle linting of internal packages like sdk, runtime, cli, etc. The benefit here is that we have a clear mental model of how linting works. All non-template packages are linted using the configuration at the root of the mono repo, and all template projects us the `eslint-config-pwa-kit` and `babel-preset-pwa-kit` pacakges. 

There are however some liberties that we took to make this happen. One of those liberties, is the use of a eslint patch package called `@rushstack/eslint-patch`, the purpose of this patch is to manipulate how dependencies are loaded in the context of a mono repo. You can read more about it [here](https://www.npmjs.com/package/@rushstack/eslint-patch). 

The second caveat is that unfortunately the above patch only works for eslint plugins, it doesn't work for 3rd party parsers (the "parser" option). What does this mean? Well we use `@babel/eslint-parser` in our eslint config, we do this because we use babel to transpile our code, so it makes sense to use babel to parse our code for linting since it'll work seamlessly with our babel config. So we have to include our new `babel-preset-pwa-kit` package as a peer dependency of the `eslint-config-pwa-kit`. This means there is a total of 3 dependencies that must be included for any of our templates to have functional linting. Our babel preset, our eslint config, and eslint itself. Which isn't too long of a list.


There have been questions on why not all packages be linted with the same set of rules. This is a good question: I do however like the separation as we might want to have distinguishing  differences like header definitions that change from one project to the next, or one project might use typescript vs ecmascript so their might have different presets of eslint configs. This the set up I think it would be weird for the mono repo root to consume a project it defines. 

Below is a visualization of what babel and eslint is responsible for what files in the mono repo. The simplest way to explain this is that, only the templates use the `babel-preset-pwa-kit` and `eslint-config-pwa-kit`. All other files that are possibly linted will be linted with the settings defined in the root `.eslintrc.js` file in conjunction with the root `babel.config.js`.

![Screen Shot 2022-04-26 at 12 49 01 PM](https://user-images.githubusercontent.com/8902581/165381609-c57c5917-674e-412c-ab03-e6f356cb0e38.png)

# Positive Side Effects

Previously we weren't linting any files in the mono repo that weren't in the pacakges folder. For example the scripts I believe weren't getting linted. This works now. 

# Other Required Changes

As a result of re-organizing the locations of the eslint configurations we have to re-solve the problem where we want copyright header enforcement for templates whilst developing in the mono-repo but not when developing a generated project (psst.. I'm not sure this worked before or not). Anyway, the solution for this is to create a new `.eslintrc.js` asset for generated projects which would essentially turn off the header rule. 

Doing this work, I modified the generator to have a "shared" folder of assets that all projects would have copied to them. You might notice that the one file in that folder is prefixed with a `$` denoting that it's a template and should be processes as one. This convention was created not only to provide some extra functionality, but to fix an issue there the `.eslintrc.js` "dot" file was getting read by eslint and breaking things. This asset shouldn't be used, and a way around it is to rename it.

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- (step1)

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

alexvuong committed:
create eslint config package

alexvuong committed:
create eslint config package

bendvc committed:
Hacking around

bendvc committed:
More hacking around, fixed eslint-parser reference

bendvc committed:
eslint command works in cli now

wjhsf commented on packages/eslint-config-pwa-kit/package.json:
```diff
@@ -0,0 +1,33 @@
+    "type": "git",
+    "url": "git+https://github.com/SalesforceCommerceCloud/pwa-kit.git"
+  },
+  "license": "SEE LICENSE IN LICENSE",
```
```suggestion
  "homepage": "https://github.com/SalesforceCommerceCloud/pwa-kit/tree/develop/packages/eslint-config-pwa-kit#readme",
  "bugs": {
    "url": "https://github.com/SalesforceCommerceCloud/pwa-kit/issues"
  },
  "license": "SEE LICENSE IN LICENSE",
```

alexvuong replied:
Do we really need README for this package 🤔 ?

wjhsf replied:
All packages need a readme!

wjhsf commented on packages/template-retail-react-app/.eslintrc.js:
```diff
@@ -6,5 +6,5 @@
 
 module.exports = {
-    extends: require.resolve('pwa-kit-cli/configs/eslint/eslint-config')
+    extends: 'pwa-kit'
```
```suggestion
    extends: ['pwa-kit']
```
Very minor, but helps new developers understand you can extend more than one config.

wjhsf commented on packages/pwa-kit-cli/bin/cli.js:
```diff
@@ -195,13 +195,8 @@ const main = () => {
-                    fix ? ' --fix' : ''
-                } "${path}"`
-            )
+            const eslint = p.join('.', 'node_modules', '.bin', 'eslint')
```
This will only work in a project root directory. Using `require.resolve` or a [find-up](https://www.npmjs.com/package/find-up) algorithm will be more resilient.

wjhsf commented on packages/pwa-kit-cli/bin/cli.js:
```diff
@@ -195,13 +195,8 @@ const main = () => {
         .argument('<path>', 'path or glob to lint')
         .option('--fix', 'Try and fix errors (default: false)')
```
We don't need to do anything special any more, so we should just pass all args straight to eslint.

Or just delete the lint command. 😉

wjhsf commented on packages/template-retail-react-app/app/components/_app/index.jsx:
```diff
@@ -55,7 +55,7 @@ const DEFAULT_NAV_DEPTH = 3
 
 const App = (props) => {
-    const {children, targetLocale, messages, categories: allCategories = {}} = props
+        const {children, targetLocale, messages, categories: allCategories = {}} = props
```
If things are configured correctly, this will cause the build to fail.

wjhsf commented on packages/eslint-config-pwa-kit/package.json:
```diff
@@ -0,0 +1,33 @@
+  "scripts": {
+    "lint": "npm run lint:js",
+    "lint:fix": "npm run lint:js -- --fix",
+    "lint:js": "internal-lib-build lint ."
```
```suggestion
    "lint:js": "eslint -c index.js ."
```

wjhsf commented on packages/eslint-config-pwa-kit/index.js:
```diff
@@ -4,8 +4,11 @@
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
+
+require("@rushstack/eslint-patch/modern-module-resolution")
```
This changes the behavior of `eslint` in ways that users might not expect, so this should be opt-in behavior. (That's what the packages [README](https://www.npmjs.com/package/@rushstack/eslint-patch#how-to-use-it=) suggests doing.)

The patch should be moved to the downstream packages with a comment explaining what it does and how to make eslint work if you don't want it (i.e. npm install [these deps](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/0efb9a53b5f90c346b850d15833032b5689f36db/packages/eslint-config-pwa-kit/package.json#L17-L29))

bendvc replied:
I have not been ignoring this comment :). This is a valid comment. I think what we can do here is to use the patch similar method to how it's described in the README (not the "lean" version used here).. and have the generator handle scrubbing out the patch.

bendvc committed:
User eslint config in other templates, add typscript support

bendvc committed:
Small changes to configs

bendvc committed:
Move lint setup to the mono-repo

bendvc committed:
fix npm scripts to use root eslint

bendvc committed:
Minor fixes

bendvc committed:
Experiment with babel preset

bendvc committed:
Remove unused babel configurations

bendvc committed:
Don't use babel config from internals that is on the chopping block

bendvc committed:
Remove old babel configs

Fix test setup
Use babel preset in webpack config

bendvc committed:
Fix lint errors

bendvc committed:
Merge branch 'v2' into eslint-work

bendvc committed:
Rebuild lock files after automatic merging

bendvc committed:
Remove debug logging

bendvc committed:
Add prettier rc files to new pacakges

bendvc committed:
Add engines config for new packages

bendvc committed:
Fix lock file

bendvc committed:
Add @loadable/babel-plugin

bendvc committed:
Merge branch 'v2' into eslint-work

bendvc committed:
Update lock files and version numbers

wjhsf commented on packages/pwa-kit-cli/bin/cli.js:
```diff
@@ -189,21 +188,6 @@ const main = () => {
-        })
-
     program
         .command('format')
```
linting and formatting usually go together. Can the `format` command also be removed?

bendvc replied:
So our eslint configurations uses the `prettier` plugin. I would agree that it's probably not required to have an additional command to only run prettier when it can be run via lint:fix. I don't know if the distinction between the two is enough to warrant a separate command.

wjhsf commented on packages/babel-preset-pwa-kit/package.json:
```diff
@@ -0,0 +1,39 @@
+    "scripts": {
+      "lint": "npm run lint:js",
+      "lint:fix": "npm run lint:js -- --fix",
+      "lint:js": "../../node_modules/.bin/eslint \"**/*.{js,jsx}\""
```
```suggestion
      "lint:js": "eslint \"**/*.{js,jsx}\""
```
Doesn't npm automatically search parent directories? Or does it stop when it finds the package.json?

bendvc replied:
It doesn't resolve executables like it does resolve dependencies. This is a temporary solution to get things working. This won't work on windows. I'm looking for alternatives atm

wjhsf commented on packages/eslint-config-pwa-kit/package.json:
```diff
@@ -0,0 +1,51 @@
+    "type": "git",
+    "url": "git+https://github.com/SalesforceCommerceCloud/pwa-kit.git"
+  },
+  "license": "SEE LICENSE IN LICENSE",
```
Don't forget to add LICENSE!

wjhsf commented on packages/internal-lib-build/bin/internal-lib-build.js:
```diff
@@ -17,16 +17,14 @@ const execSync = (cmd, opts) => {
     program.command('build').action(() => {
         const babel = p.join(binDir, 'babel')
-        const babelConfig = p.resolve(p.join(process.cwd(), 'babel.config.js'))
+        const babelConfig = p.resolve(p.join(process.cwd(), '..', '..', 'babel.config.js'))
         const prepareDist = p.resolve(p.join(__dirname, 'prepare-dist.js'))
```
```suggestion
        const babelConfig = p.join(__dirname, '..', '..', 'babel.config.js')
        const prepareDist = p.join(__dirname, 'prepare-dist.js')
```
If you're looking for a specific config in a specific place, you probably shouldn't use `process.cwd()`. (Also `resolve` is unnecessary here.)

wjhsf commented on packages/babel-preset-pwa-kit/index.js:
```diff
@@ -0,0 +1,43 @@
+                require('@babel/preset-env'),
+                {
+                    targets: {
+                        node: 12
```
node 12 is end-of-life at the end of the week. Do we still want to target it or should we bump to 14?

bendvc replied:
Good call out. It would be nice to keep that change in a separate pr. This one is getting a little large as it is.

wjhsf commented on packages/babel-preset-pwa-kit/index.js:
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
```
```suggestion
 * Copyright (c) 2022, Salesforce, Inc.
```

wjhsf commented on .eslintrc.js:
```diff
@@ -1,6 +1,40 @@
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
```
```suggestion
 * Copyright (c) 2022, Salesforce, Inc.
```

wjhsf commented on packages/template-retail-react-app/package.json:
```diff
@@ -28,6 +29,8 @@
         "core-js": "2.4.0",
         "cross-env": "^5.2.0",
         "cross-fetch": "^3.1.5",
+        "eslint": "7.31.0",
```
```suggestion
        "eslint": "^7.31.0",
```

wjhsf commented on packages/template-typescript-minimal/app/pages/home.tsx:
```diff
@@ -11,7 +11,7 @@ import HelloTS from '../components/hello-typescript'
 
 interface Props {
-    dadJoke: string
+    dadJoke: any
```
Why this change?

alexvuong replied:
the dadJoke from the api is an object, not a string. More precisely, we can declare its interface. 
```
interface dadJoke {
  id: string,
  joke: string,
  status: number. 
}
```

wjhsf commented on scripts/smoke-test-npm-scripts.js:
```diff
@@ -1,4 +1,10 @@
 #!/usr/bin/env node
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
```
```suggestion
 * Copyright (c) 2021, Salesforce, Inc.
```

wjhsf commented on scripts/gtime.js:
```diff
@@ -1,4 +1,10 @@
 #!/usr/bin/env node
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
```
```suggestion
 * Copyright (c) 2022, Salesforce, Inc.
```

wjhsf commented on scripts/count-dependencies.js:
```diff
@@ -1,4 +1,10 @@
 #!/usr/bin/env node
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
```
```suggestion
 * Copyright (c) 2021, Salesforce, Inc.
```

wjhsf commented on scripts/check-version.js:
```diff
@@ -1,4 +1,10 @@
 #!/usr/bin/env node
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
```
```suggestion
 * Copyright (c) 2022, Salesforce, Inc.
```

wjhsf commented on scripts/check-dependencies.js:
```diff
@@ -1,4 +1,10 @@
 #!/usr/bin/env node
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
```
```suggestion
 * Copyright (c) 2022, Salesforce, Inc.
```

wjhsf commented on scripts/bump-version.js:
```diff
@@ -1,4 +1,10 @@
 #!/usr/bin/env node
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
```
```suggestion
 * Copyright (c) 2022, Salesforce, Inc.
```

wjhsf commented on scripts/bootstrap.js:
```diff
@@ -1,4 +1,10 @@
 #!/usr/bin/env node
+/*
+ * Copyright (c) 2021, salesforce.com, inc.
```
```suggestion
 * Copyright (c) 2022, Salesforce, Inc.
```

wjhsf commented on packages/template-retail-react-app/babel.config.js:
```diff
@@ -4,4 +4,9 @@
  */
-module.exports = require('pwa-kit-cli/configs/babel/babel-config')
+
+// NOTE: THIS IS AN EXAMPLE OF HOW TO USE A CUSTOM BABEL CONFIGURATION. SIMPLY DEFINE A BABEL.CONFIG.JS
+// AND USE THE PWA-KIT PRESET ADDING ANY ADDITIONAL PLUGINS OR CHANGES NECESSARY.
```
```suggestion
// NOTE: THIS IS AN EXAMPLE OF HOW TO USE A CUSTOM BABEL CONFIGURATION. SIMPLY DEFINE A BABEL.CONFIG.JS
// AND USE THE PWA-KIT PRESET, ADDING ANY ADDITIONAL PLUGINS OR CHANGES NECESSARY.
```
Mostly the comma, but also all caps is rather aggressive.
```suggestion
// Note: This is an example of how to use a custom babel configuration. Simply define a babel.config.js
// and use the pwa-kit preset, adding any additional plugins or changes necessary.
```

bendvc replied:
This isn't something that I'm planning on leaving in the code. It was solely for demo purposes.

wjhsf reviewed (commented):
> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this comment are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

All files in the monorepo MUST have a copyright header. All files in generated projects SHOULD NOT enforce a header. With these changes, how is this accomplished?

bendvc committed:
Fix babel-preset dependency for npm 7

Remove `enzyme-adapter-react-16` as it's not used.

bendvc committed:
Fix lint issues in generator assets

bendvc committed:
Use asset for turn off header config for generated projects

bendvc committed:
We expect the output dir to be returned.

bendvc committed:
Fix copy command

bendvc committed:
debugging cp command

bendvc committed:
Another guess fix

bendvc committed:
Update create-mobify-app.js

bendvc committed:
try 1 file?

bendvc commented:
> > The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this comment are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).
> 
> All files in the monorepo MUST have a copyright header. All files in generated projects SHOULD NOT enforce a header. With these changes, how is this accomplished?

So the way this is working now is that during the generation step, the `.eslintrc.js` will be replaced with one where the `header/header` rule is turned off. Similar to this:

```
module.exports = {
    root: true,
    extends: ['pwa-kit'],
    rules: {
        'header/header': 'off'
    }
}
```

The partner dev's as choose to leave it like that, or to define their own header rule and run with it. The only issue is that the generator (create-app package) doesn't have the best implementation of dealing with assets, so although I can get this working quickly it would be dirty, and needs some refactoring.

bendvc committed:
Move eslint asset into "shared" folder

bendvc committed:
Update create-mobify-app.js

bendvc committed:
Update create-mobify-app.js

bendvc committed:
Fix windows linting, use template for eslint rc in generator

bendvc committed:
Support template and file copy

bendvc committed:
Merge branch 'v2' into eslint-work

bendvc committed:
Update lock files

bendvc committed:
Ensure output dir is available after all presets

bendvc committed:
Add licences, readmes and changelogs

bendvc committed:
Remove retail react app babel config

not required as it is the default implementation

bendvc committed:
Fix copy headers for scripts

bendvc committed:
PR feedback

bendvc committed:
Update packages/eslint-config-pwa-kit/package.json

Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>

bendvc committed:
Merge branch 'eslint-work' of https://github.com/SalesforceCommerceCloud/pwa-kit into eslint-work

bendvc committed:
Update readmes

bendvc committed:
Update .eslintrc.js

Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>

bendvc committed:
Update packages/babel-preset-pwa-kit/index.js

Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>

bendvc committed:
Fix copyrite

bendvc committed:
Pass answers to shared file/template copy

bendvc committed:
Only use pwa-kit preset if no babel config is defined

bendvc committed:
Merge branch 'eslint-work' of https://github.com/SalesforceCommerceCloud/pwa-kit into eslint-work

wjhsf commented:
I don't remember if you mentioned this, but I noticed that linting in VS Code does not work for `template-*` packages. It can't find `babel-preset-pwa-kit`, which seems to be because it's only looking in the project root.

wjhsf commented on packages/pwa-kit-create-app/assets/shared/$.eslintrc.js:
```diff
@@ -4,4 +4,12 @@
+    extends: ['pwa-kit'],
+    rules: {
+        'header/header': 'off'
+    }
```
If we have to have two separate configurations for in the monorepo and in generated packages, then I think we should flip this. The header rule should not exist in the base eslint config that we publish; it is specific to our use case. That way users of the config can just use it out of the box, and don't need to do any customizations to disable the header that none of them will want.

wjhsf replied:
I think that the cleanest way to have the header work inside the monorepo but not outside is to have it defined in the monorepo root and avoid using `root: true`. However, given all of the other constraints we're facing, I'm not sure how viable that is as an option.

bendvc committed:
Merge branch 'v2' into eslint-work

bendvc committed:
Re-generate lock files

bendvc committed:
Reinstate jest setup, removed in merge

bendvc committed:
Fix jest setup

bendvc committed:
Merge branch 'v2' into eslint-work

bendvc committed:
Fix version numbers

bendvc committed:
Re generate lock files after merge of bumped version

adamraya commented on packages/babel-preset-pwa-kit/package.json:
```diff
@@ -0,0 +1,44 @@
+      "lint:fix": "npm run lint:js -- --fix",
+      "lint:js": "cross-env ../../node_modules/.bin/eslint \"**/*.{js,jsx}\""
+    },
+    "dependencies": {
```
We need to include these two plugins as dependencies:
https://github.com/SalesforceCommerceCloud/pwa-kit/blob/88ad67fbfdccb9f19b33681df2c5fa4ef07e0d7d/packages/babel-preset-pwa-kit/index.js#L25-L26

adamraya commented on packages/babel-preset-pwa-kit/package.json:
```diff
@@ -0,0 +1,44 @@
+    "dependencies": {
+        "@babel/plugin-proposal-object-rest-spread": "^7.4.4",
+        "@babel/plugin-transform-async-to-generator": "^7.7.0",
+        "@babel/plugin-transform-modules-commonjs": "^7.4.4",
```
We're not using the `@babel/plugin-transform-modules-commonjs` plugin in the `babel-preset-pwa-kit` package. 
Do we need to include it as a dependency?

alexvuong commented on packages/eslint-config-pwa-kit/package.json:
```diff
@@ -0,0 +1,58 @@
+    "@babel/eslint-parser": "^7.17.0",
+    "@rushstack/eslint-patch": "^1.1.3",
+    "@typescript-eslint/parser": "^5.20.0",
+    "babel-preset-pwa-kit": "^2.0.0-dev.6",
```
Do we need it as dependency here? 🤔

alexvuong replied:
Reason I asked is if `babel-preset-pwa-kit` is a dependency of `eslint-config`. When a generated project install the `eslint-config`, the babel preset packages should also be installed as well. 

With that logic, do we still need to have `babel-preset-pwa-kit` to be listed as a dependency in each template-*?

alexvuong replied:
To verify the eslint still works without explicitly having `babel-preset-pwa-kit`. what I did:
-  Removed the `babel-preset-pwa-kit` [here](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/529/files#diff-00e24db644be6a5f62f06ee3d5eb644005bdb9590ebc8ddf6a3c03d1b82fa6ebR25) from `template-retail-react-app`. 
- Generated the project with command 
` GENERATOR_PRESET=test-project node packages/pwa-kit-create-app/scripts/create-mobify-app-dev.js --outputDir ../generated-project-no-babel`
- Opened the generated project with my IDE, the eslint works as expected without throwing any issue
![image](https://user-images.githubusercontent.com/52219283/166352946-bb211c8e-4ac2-491e-96f2-d12d95a17758.png).

From this, either we
- Have `babel-preset-pwa-kit` as dependency, and do not need to install it inside `template-*`
- Do not include it as dependency here, only peerDependency, and have it as dependencies in template-*

alexvuong commented on package.json:
```diff
@@ -6,10 +6,38 @@
+    "eslint-plugin-jsx-a11y": "6.4.1",
+    "eslint-plugin-prettier": "3.0.1",
+    "eslint-plugin-react": "7.24.0",
+    "prettier": "^1.18.2",
```
With the proposed solution, I suppose now we have two sets of eslint we need to maintain, is there a necessity that we keep both of them to be the same?

vmarta replied:
@alexvuong I believe that we want to have multiple setups: one for the monorepo developers and the other one for other developers working on their pwa-kit projects. Even though both setups start out to be nearly identical, but they may deviate more over time. So I think the duplicate code is all right. 

But @bendvc you would know better. Was it intended to be that way?

wjhsf replied:
If we change the root eslint config to `{ extends: ['pwa-kit'] }` and add `./packages/eslint-config-pwa-kit` and `./packages/babel-preset-pwa-kit` as dev dependencies then we will only need to maintain the shared config values (internal + external) in a single location.

bendvc replied:
@vmarta Yes, you are correct, that is the intention. That is also what we had prior to this change. You can see that there are eslint configurations in both the "cli" and the "internal build tools". One of the issues with the eslinting before this was that only projects get linted, not the scripts folder that is a sibling folder. I thought that was less than ideal. 

@wjhsf Yes you can do something like that. Do I think it's a little weird to use the output of a package of a mono repo in the root, maybe.

adamraya reviewed (commented):
Optional Nit.
> I modified the generator to have a "shared" folder of assets that all projects would have copied to them.

I believe this is true for the whole `packages/pwa-kit-create-app/assets` folder.

I propose:
1. Rename.
`packages/pwa-kit-create-app/assets/pwa` → `packages/pwa-kit-create-app/assets/config`

2. Remove the `shared` folder.
`packages/pwa-kit-create-app/assets/shared/$.eslintrc.js` → `packages/pwa-kit-create-app/assets/$.eslintrc.js`

 3. Follow the pattern of adding template files to the `packages/pwa-kit-create-app/assets/` that the generator adapts to the generated project, not files that we copy directly to a generated project.

vmarta commented on packages/pwa-kit-create-app/package.json:
```diff
@@ -22,7 +22,7 @@
     "lint": "npm run lint:js",
     "lint:fix": "npm run lint:js -- --fix",
-    "lint:js": "internal-lib-build lint .",
+    "lint:js": "cross-env ../../node_modules/.bin/eslint \"**/*.{js,jsx}\"",
```
Just curious: in this case, for what purpose is `cross-env` being used for?

bendvc replied:
In this case cross-env will resolve the forward-slashes to it makes it work on windows machines.

vmarta commented on packages/template-express-minimal/.eslintrc.js:
```diff
@@ -5,5 +5,6 @@
  */
 module.exports = {
-    extends: require.resolve('pwa-kit-cli/configs/eslint/eslint-config')
+    root: true,
```
TIL about the cascading behaviour of ESLint config

adamraya commented on packages/pwa-kit-react-sdk/package.json:
```diff
@@ -26,7 +26,7 @@
     "lint": "npm run lint:js",
     "lint:fix": "npm run lint:js -- --fix",
-    "lint:js": "internal-lib-build lint \"**/*.{js,jsx}\"",
+    "lint:js": "cross-env ../../node_modules/.bin/eslint \"**/*.{js,jsx}\"",
```
Instead of assuming that we'll find an `eslint` binary here `../../node_modules/.bin/eslint` looks like we should add eslint as a `devDependency` and use it in a script like `"lint:js": "eslint ..."` as a shortcut for the local `node_modules/.bin/eslint`.

I think this could fail if, for example, someone tries to use the `pwa-kit-react-sdk` package outside of the monorepo and the project doesn't have the eslint bin in that folder.

vmarta commented on .eslintignore:
```diff
@@ -0,0 +1,4 @@
+node_modules
+template-express-minimal/**
+template-retail-react-app/**
+template-typescript-minimal/**
```
These template paths were not matching, since glob pattern like `/**` is relative to the location of the .eslintignore file. It's a subtle difference when compared to passing in just `node_modules`. For more details, see [.gitignore spec](https://git-scm.com/docs/gitignore).

So what would work is: `packages/template-express-minimal/**`. But since we would have more templates in the future, how about this instead?

```suggestion
packages/template-*/**
```

bendvc replied:
Excellent. I was going to get around to do that. Looks like you beat me to it. 

NOTE: I've recently been forced to deal with more globs, [this](https://www.digitalocean.com/community/tools/glob?comments=true&glob=packages%2Ftemplate-%2A%2F%2A%2A&matches=false&tests=packages%2Ftemplate-express-minimal%2Ftest%2Fblah.ts&tests=packages%2Ftemplate-retail-react-app%2Ffile.js&tests=packages%2Ftemplate-typescript-minimal) tool helps when testing.

vmarta commented:
Regarding **VSCode's settings.json** mentioned in the PR description: I feel like by itself, it can open up possibilities. For example, the original issues were both eslint (and its plugins) and its config file cannot be found. With VSCode's settings.json, you can configure it to find all of the eslint-related stuffs:

```json
  "eslint.options": {"resolvePluginsRelativeTo": "./packages/pwa-kit-cli"},
  "eslint.nodePath": "./packages/pwa-kit-cli/node_modules",
```

bendvc commented:
> Optional Nit.
> 
> > I modified the generator to have a "shared" folder of assets that all projects would have copied to them.
> 
> I believe this is true for the whole `packages/pwa-kit-create-app/assets` folder.
> 
> I propose:
> 
> 1. Rename.
>    `packages/pwa-kit-create-app/assets/pwa` → `packages/pwa-kit-create-app/assets/config`
> 2. Remove the `shared` folder.
>    `packages/pwa-kit-create-app/assets/shared/$.eslintrc.js` → `packages/pwa-kit-create-app/assets/$.eslintrc.js`
> 3. Follow the pattern of adding template files to the `packages/pwa-kit-create-app/assets/` that the generator adapts to the generated project, not files that we copy directly to a generated project.

So I've had discussions with @olibrook about how we might make changes to the create-app specifically how the assets/templated files work. There are alternatives to what you suggested, and we can talk about those later. But the thing I want to get across is that it's probably a larger set of work outside of the scope of this project. Luckily we can most likely make changes to the generator without breaking anything.

bendvc commented:
> Regarding **VSCode's settings.json** mentioned in the PR description: I feel like by itself, it can open up possibilities. For example, the original issues were both eslint (and its plugins) and its config file cannot be found. With VSCode's settings.json, you can configure it to find all of the eslint-related stuffs:
> 
> ```json
>   "eslint.options": {"resolvePluginsRelativeTo": "./packages/pwa-kit-cli"},
>   "eslint.nodePath": "./packages/pwa-kit-cli/node_modules",
> ```

@vmarta are you suggesting that a potential solution to the original problem could be solved by providing a .vscode/settings.js file to users? Or some kind of help doc. As long as the lint commands work in the cli we are ok?

bendvc committed:
Update .eslintignore

Co-authored-by: Vincent Marta <vmarta@salesforce.com>

vmarta commented:
> > Regarding **VSCode's settings.json** mentioned in the PR description: I feel like by itself, it can open up possibilities. For example, the original issues were both eslint (and its plugins) and its config file cannot be found. With VSCode's settings.json, you can configure it to find all of the eslint-related stuffs:
> > ```json
> >   "eslint.options": {"resolvePluginsRelativeTo": "./packages/pwa-kit-cli"},
> >   "eslint.nodePath": "./packages/pwa-kit-cli/node_modules",
> > ```
> 
> @vmarta are you suggesting that a potential solution to the original problem could be solved by providing a .vscode/settings.js file to users? Or some kind of help doc. As long as the lint commands work in the cli we are ok?

@bendvc yeah, it's an idea that came up, since we already consider to bring in vscode settings file. We can provide:
- vscode workspace **settings file** - for developers who use vscode
- and like you mentioned, a **help doc** to provide guidance for developers using other text editors (WebStorm, vim). I think those editors would allow you to do similar manual configuration of eslint.

Just playing around, here's what the vscode settings might look like: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/583

wjhsf committed:
Who doesn't love more ESLint? (#579)

* hack: get `lerna bootstrap` working

* add linting errors for testing

Observation: CLI linting works across the board
Observation: IDE linting broken in template-* packages
Cannot find module 'babel-preset-pwa-kit'

* Fix VS Code / ESLint plugin not working.

* Do some rekerjiggering.

* Update deps.

* Restore eslint config.

* Fix eslint scripts.

* Remove header rule.

* Remove generated eslint config.

* Revert package-locks to lockfileVersion 1.

* gitignore .vscode/ except settings.json

* Bump to latest prerelease.

* Remove self-dependency.

* Fix duplicate headers bug.

More info: https://github.com/Stuk/eslint-plugin-header/issues/39

* Make VS Code lint *all* the files!

* Fix linter errors.

* Remove unused cross-env.

* Fix linter error.

* Don't lint babel config :(

* Add linting for `scripts` directory.

bendvc committed:
Remove scripts lint config

This was only working for the IDE (sometimes) and not for the cli, which was not in alignment how the rest of the packages, worked

bendvc committed:
Fix babel for jest tests

bendvc commented:
We are abandoning the ESLint re-work. The final nail in the coffin so to speak is that solution fails to meet the original acceptance criteria of making IDE integration work. Although the architecture is more ideal, recent large changes had made for a complex and difficult merge conflict with `v2`. So much so that we won't be able to resolve it before the upcoming release. 

This branch is in a bad state, if we were to bick up work later I would suggest doing so from this commit 👉  a9ab78a39fdd973061f75f411b2f72e670b05ee5.
